SourceX/miniwin/dsound.cpp

```diff
@@ -48,7 +48,14 @@ ULONG DirectSoundBuffer::Release()
  */
 HRESULT DirectSoundBuffer::GetStatus(LPDWORD pdwStatus)
 {
-	return DVL_DSERR_INVALIDPARAM;
+	for (int i = 1; i < Mix_AllocateChannels(-1); i++) {
```

Comment:
Why not just `switch (SDL_GetAudioStatus())`?

Reply:
Wouldn't `SDL_GetAudioStatus()` almost always return `SDL_AUDIO_PLAYING` because of the music?

From my understanding, `GetStatus()` should return the state of its own `DirectSoundBuffer` instance, not the whole device.

Reply:
Ah, I think you're right! Thanks for explaining.

Reply:
@Pyr0Byt3 looks like your right, SDL_GetAudioStatus is on the device level and it might even return true when Mixer is just playing silence on all channels of that device. From what I can gather the way you did it is the correct way to check if a sound is playing. But then we probably have to implement DuplicateSoundBuffer for making sure that multiple instances can be played when that is needed. It feels like the semantics between SDL and DSound to map so well for this case.

Reply:
`Play()` uses `Mix_PlayChannel(-1)`, which lets multiple sounds play (it uses the first available channel it finds).
https://github.com/diasurgical/devilutionX/blob/3de9b4b1e96a52196d6150a220b106620dd467c5/SourceX/miniwin/dsound.cpp#L63-L65

I think a more correct abstraction would be to keep track of `channel` in a private field, that way `GetStatus()` and `Stop()` wouldn't need to loop across all channels looking for the same `chunk`. But that's kind of unnecessary if you plan on replacing all the DirectSound stuff with SDL anyway.

The only oddity resulting from using `Mix_PlayChannel(-1)` instead of `DuplicateSoundBuffer()` (that I can think of) is that the buffers in `sound.cpp` go unused:
https://github.com/diasurgical/devilutionX/blob/3de9b4b1e96a52196d6150a220b106620dd467c5/SourceX/sound.cpp#L8

So more than 8 sounds could play at the same time.

Reply:
Agreed, the original limit would be n+8 (where n is the number of wav files in Diablo). Lets not do any extra work in order to implement the original limitations, 8 was probably just a number that they felt was a fair between scenarios and memory constraints.

I did some tests in Lazarus' chamber (with unlimited duplicate sounds) and settled on 25 channels as being fairly close to the max number that you would normally play at any one time (I have seen higher in a few instances).

This is the best info I have been able to find regarding the DirectSound limit, it's from Wikipedia, but I believe I originally saw something very similar on microsoft.com (sans the performance caveat).
> DirectSound provides sample rate conversion and sound mixing (volume and pan) for an unlimited number of audio sources; however, the practical limits are the number of hardware audio sources and the performance of software mixers.

As such I don't think we could ever be sure to match the original behavior unless we allocate a channel for all loaded sounds and that just seams a bit over eager :)